apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,10 +1428,231 @@ private void StartPush()
                 }
             }
 
+            /// <summary>
+            /// This method is called when the acknowledgement request completes.
+            /// This method handles the response from an acknowledgement request for exactly once subscriptions only.
+            /// For non-exactly once subscriptions, <see cref="HandleAckResponse(Task, List{string}, List{string}, List{TimedId})"/> is called.
+            /// </summary>
+            /// <param name="writeTask">The task containing the response of Acknowledge/ModifyAckDeadline RPC.</param>
+            /// <param name="ackIds">The list of ids that were sent to the Acknowledge RPC for acknowledgement
+            /// that the message associated with these ids shouldn't be redelivered.</param>
+            /// <param name="nackIds">The list of ids that were sent to the ModifyAckDeadline RPC to indicate that
+            /// the message associated with these ids should be redelivered.</param>
+            /// <param name="extendIds">The list of ids that were sent to the ModifyAckDeadline RPC to modify their ack deadline.</param>
+            /// <remarks>
+            /// Only one of ackIds, nackIds and extendIds will be non-null and based on that we determine if we are handling acks, nacks or extends.
+            /// This method is used for exactly once subscriptions only. To keep the logic of exactly once subscription in one place,
+            /// this method has a few local methods that are used only in exactly once flow and not in non-exactly once flow, so the method
+            /// is pretty long.
+            /// </remarks>
+            private void HandleAckResponseForExactlyOnce(Task writeTask, List<string> ackIds, List<string> nackIds, List<TimedId> extendIds)
+            {
+                _concurrentPushCount -= 1;
+                _pushInFlight -= (ackIds?.Count ?? 0) + (nackIds?.Count ?? 0) + (extendIds?.Count ?? 0);
+                
+                bool hasAckIds = ackIds?.Count > 0;
+                bool hasNackIds = nackIds?.Count > 0;
+                bool hasExtendIds = extendIds?.Count > 0;
+                // In one call, we'll have one of acks or nacks or extends, not all.
```

Comment:
Consider an assertion for this.

Reply:
We have plenty of precondition checks around the libraries - mostly for externally-provided arguments, but it would be fine to use here. I'd suggest using `GaxPrecondition.CheckArgument`. Admittedly I expect the expression will be slightly fiddly to write - something like:

```csharp
// Any of the parameter names would be okay here, I think...
GaxPreconditions.CheckArgument((hasAckIds ? 1 : 0) + (hasNackIds ? 1 : 0) + (hasExtendIds ? 1 : 0) <= 1, nameof(...), "At most one of AckIds, NackIds or ExtendIds should be non-empty.");
```

Reply:
> We have plenty of precondition checks around the libraries - mostly for externally-provided arguments, but it would be fine to use here. I'd suggest using `GaxPrecondition.CheckArgument`. Admittedly I expect the expression will be slightly fiddly to write - something like:
> 
> ```cs
> // Any of the parameter names would be okay here, I think...
> GaxPreconditions.CheckArgument((hasAckIds ? 1 : 0) + (hasNackIds ? 1 : 0) + (hasExtendIds ? 1 : 0) <= 1, nameof(...), "At most one of AckIds, NackIds or ExtendIds should be non-empty.");
> ```

Thanks Jon. Somehow, immediately after hearing assertion, `Debug.Assert` , `Trace.Assert` came to my mind. I will now add your code suggestion.